### DISCUSSION TITLE
Frequent duplicate API calls being processed

### DISCUSSION THREAD

**Discussion by Ben-CA** [socketio/socket.io#4357] [Q&A]
I'm frequently getting multiple of the same API calls being made that have been passed through socket.IO. I have client-side code (in React) that emits a command:

```
  const issueMaterial = (workOrderNumber, itemId, recordId, quantity) => {
    socket.emit('Issue Material', {
      token,
      workOrderNumber,
      itemId,
      recordId,
      quantity,
      socketId: socket.id,
    })
  }
```

And then I have Node.js server-side code that receives the emitted command, and puts it into a queue to be processed by the API:

```
const socketIo = require('socket.io')
const { webApiProducer } = require('../queues/webapi.queue')

class SocketService {
  constructor(server) {
    this.io = socketIo(server, {
      cors: {
        origin: '*',
      },
    })
    this.io.on('connection', (socket) => {
      console.log(`Web Socket User Connected with ID: ${socket.id}`)

      socket.on('Issue Material', (message) => {
        webApiProducer({ ...message, type: 'Issue Material' })
      })
    })
  }

  emiter(event, body, socketId) {
    if (body) this.io.to(socketId).emit(event, body)
  }
}

module.exports = SocketService
```

Not always, but I'm often 'randomly' getting duplicates / triplicates, etc. Do I need to be using socket.off() or socket.removeAllListeners() somewhere in my code to avoid duplicates?

**Comment by darrachequesne** [comment]
Hi! Could it be caused by multiple clicks on the client side? You should be able to temporarily disable the button when the user clicks on it, to prevent double clicks.

**Reply by Ben-CA** [reply]
That's originally what I thought it might be, but I've disabled the button once it's clicked, and even when I try "rage clicking" to get a "double click" in testing, it won't double trigger.

**Reply by Ben-CA** [reply]
Still trying to get to the bottom of this, but it seems it's not actually an issue with my Socket setup.
